frontends/p4/structInitializers.cpp

```diff
@@ -13,10 +13,166 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+                    auto stat = new IR::MethodCallStatement(mc->srcInfo, mc);
+                    retval->push_back(stat);
+                }
+            } else if (type->is<IR::Type_BaseList>()) {
+                vec->push_back(resolveDefaultInitTuple(expr, type, retval, refMap));
+            } else if (type->is<IR::Type_Struct>()) {
+                vec->push_back(resolveDefaultInitTuple(expr, type, retval, refMap));
+            } else {
+                vec->push_back(expr);
+            }
+        }
+        auto result = new IR::ListExpression(statement->srcInfo, *vec);
+        return result;
+    } else if (auto strct = ltype->to<IR::Type_Struct>()) {
+        auto sli = statement->to<IR::StructExpression>();
+        auto si = new IR::IndexedVector<IR::NamedExpression>();
+        for (auto f : strct->fields) {
+            auto ne = sli->components.getDeclaration<IR::NamedExpression>(f->name.name);
+            if ((f->type->is<IR::Type_Header>() &&
+                 ne->expression->to<IR::StructExpression>()->defaultInitializer)
+                 || f->type->is<IR::Type_HeaderUnion>()) {
+                cstring name = refMap->newName("hdr_or_hu");
+                auto decl = new IR::Declaration_Variable(statement->srcInfo, IR::ID(name), f->type);
+                IR::Expression* pe = new IR::PathExpression(statement->srcInfo,
+                                                            f->type, new IR::Path(name));
+                ne = new IR::NamedExpression(statement->srcInfo, f->name, pe);
+                retval->push_back(decl);
+            } else if (f->type->is<IR::Type_BaseList>() || f->type->is<IR::Type_Struct>()) {
+                ne = new IR::NamedExpression(statement->srcInfo, f->name,
+                     resolveDefaultInitTuple(ne->expression, f->type, retval, refMap));
+            }
+            si->push_back(ne);
+        }
+        auto type = strct->getP4Type()->to<IR::Type_Name>();
+        return new IR::StructExpression(statement->srcInfo, type, type, *si);
+    }
+    return nullptr;
+}
+
+
+const IR::Statement* resolveDefaultInit(IR::AssignmentStatement* statement, const IR::Type* ltype,
+                                        IR::IndexedVector<IR::StatOrDecl>* retval,
+                                        ReferenceMap* refMap) {
+    auto strct = ltype->to<IR::Type_StructLike>();
+    if (auto si = statement->right->to<IR::StructExpression>()) {
+        for (auto f : strct->fields) {
+            auto right = si->components.getDeclaration<IR::NamedExpression>(f->name);
+            auto left = new IR::Member(statement->left, f->name);
+            if (f->type->is<IR::Type_Header>()) {
+                auto le = right->expression->to<IR::ListExpression>();
+                auto se = right->expression->to<IR::StructExpression>();
+                if ((le && le->defaultInitializer) ||
+                    (se && se->defaultInitializer)) {
```

Comment:
much of this code seems to be common with the previous function (not 100% clear at a cursory reading); could you refactor some of it?